File: LeetCode/cpp/93.cpp
class Solution {
public:
    vector<string> restoreIpAddresses(string s) {
        int n = s.size();
        vector<string> ret;
        vector<int> temp;
        BackTracking(s, n, 0, temp, ret);
        return ret;
    }

    void BackTracking(string &s, int &n, int i, vector<int> &temp, vector<string> &ret) {
        if (temp.size() == 4) {
            if (i == n)
                push(temp, ret);
            return;
        }
        for (int j = i; j < i + 3; ++j) {
            if (j >= n) return;
            if (stoi(s.substr(i, j - i + 1)) > 255) return;
            if (j - i > 0 && s[i] == '0') return;
            temp.push_back(stoi(s.substr(i, j - i + 1)));
            BackTracking(s, n, j + 1, temp, ret);
            temp.pop_back();
        }
    }

    void push(vector<int> &temp, vector<string> &ret) {
        string s;
        for (int i = 0; i < 4; ++i)
            s += (to_string(temp[i]) + ".");
        s.pop_back();
        ret.push_back(s);
    }
};

File: LeetCode/python3/93.py
class Solution:
    def restoreIpAddresses(self, s: str) -> List[str]:
        n = len(s)

        def Track(num, pos, str):
            if pos == n and num < 4:
                return
            if num == 4:
                if pos == n:
                    ret.append(str[:-1])
                return
            for i in range(3):
                if pos + i >= n:
                    break
                temp = s[pos: pos + i + 1]
                if (int(temp) == 0 and len(temp) == 1) or (int(temp) > 0 and int(temp) <= 255 and temp[0] != '0'):
                    str += temp + '.'
                    Track(num + 1, pos + i + 1, str)
                    str = str[:-(len(temp) + 1)]

        ret = []
        Track(0, 0, "")
        return ret
